src/Components/Popover/Popover.tsx

```diff
@@ -0,0 +1,476 @@
+
+interface Props extends StandardProps {
+  /** Popover dialog color */
+  color: ComponentColor
+  /** Popover dialog contents */
+  contents: (onHide?: () => void) => ReactChild
+  /** Type of interaction to show the popover dialog */
+  showEvent: PopoverInteraction
+  /** Type of interaction to hide the popover dialog */
+  hideEvent: PopoverInteraction
+  /** Pixel distance between trigger and popover dialog */
+  distanceFromTrigger: number
+  /** Where to position the popover relative to the trigger (assuming it fits there) */
+  position: PopoverPosition
+  /** Means of applying color to popover */
+  type: PopoverType
+  /** Renders the popover dialog visible initially */
+  initiallyVisible: boolean
+}
+
+interface State {
+  expanded: boolean
+  triggerRect: ClientRect | DOMRect | null
+}
+
+interface PopoverFlush {
+  first: boolean
+  last: boolean
+}
+
+export class Popover extends Component<Props, State> {
+  public static readonly displayName = 'Popover'
+
+  public static defaultProps = {
+    color: ComponentColor.Primary,
+    testID: 'popover',
+    showEvent: PopoverInteraction.Click,
+    hideEvent: PopoverInteraction.Click,
+    distanceFromTrigger: 4,
+    position: PopoverPosition.Below,
+    type: PopoverType.Solid,
+    initiallyVisible: false,
+  }
+
+  public static DismissButton = PopoverDismissButton
+
+  private triggerRef = createRef<HTMLDivElement>()
+  private dialogRef = createRef<HTMLDivElement>()
+
+  constructor(props: Props) {
+    super(props)
+
+    this.state = {
+      expanded: this.props.initiallyVisible,
+      triggerRect: null,
+    }
+  }
+
+  public componentDidMount() {
+    if (this.props.initiallyVisible) {
+      this.handleShowDialog()
+    } else {
+      this.handleUpdateTriggerRect()
+    }
+  }
+
+  public render() {
+    const {testID, children, id} = this.props
+
+    return (
+      <ClickOutside onClickOutside={this.handleClickOutside}>
+        <div
+          className={this.className}
+          data-testid={testID}
+          id={id}
+          onClick={this.handleTriggerClick}
+          onMouseOver={this.handleTriggerMouseOver}
+          onMouseOut={this.handleTriggerMouseOut}
+          ref={this.triggerRef}
+        >
+          {children}
+          {this.dialog}
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private get className(): string {
+    const {color, className, type} = this.props
+
+    return classnames('cf-popover', {
+      [`cf-popover__${color}`]: color,
+      [`${className}`]: className,
+      [`cf-popover__${type}`]: type,
+    })
+  }
+
+  private get dialog(): JSX.Element {
+    const {triggerRect, expanded} = this.state
+    const {contents, distanceFromTrigger} = this.props
+    const dialogPosition = this.calculateDialogPosition()
+    let dialogStyles: CSSProperties = {
+      visibility: expanded ? 'visible' : 'hidden',
+    }
+    let caretStyles: CSSProperties = {}
+
+    if (triggerRect && this.dialogRef.current) {
```

Comment:
It seems like this needs `dialogRef.current` to render, but `dialogRef.current` isn't attached until it renders. So it requires two renders to work, which may be happening incidentally. But it's hard to tell. Will this work if it is rendered exactly once?

One other way to do it is to render the dialog with its contents (which attaches the ref), but then to compute the styles in `componentDidUpdate` and attach them imperatively (i.e. `this.dialogRef.current.setAtribute('style', ...)`).